contrib/tools/workflowcheck/determinism/checker.go

```diff
@@ -0,0 +1,485 @@
+								}
+							}
+						}
+					}
+				}
+			}
+		}
+	}
+
+	// Build collector and do initial pass for each function async
+
+	// Parallelize to the number of CPUs
+	maxAtOnce := runtime.NumCPU()
+	if maxAtOnce < 1 {
+		maxAtOnce = 1
+	}
+	doneCh := make(chan struct{}, maxAtOnce)
+	var running int
+	for fn, decl := range funcDecls {
+		// If we've filled the channel, wait
+		if running == cap(doneCh) {
+			<-doneCh
+		} else {
+			running++
+		}
+		go func(fn *types.Func, decl *ast.FuncDecl) {
+			coll.collectFuncInfo(fn, decl)
+			doneCh <- struct{}{}
+		}(fn, decl)
+	}
+	// Wait for the rest to finish
+	for i := 0; i < running; i++ {
+		<-doneCh
+	}
+	// Build facts in second pass
+	return coll.applyFacts(), nil
+}
+
+func updateIgnoreMap(fset *token.FileSet, f *ast.File, m map[ast.Node]struct{}) {
+	// Collect only the ignore comments
+	var comments []*ast.CommentGroup
+	for _, group := range f.Comments {
+		if len(group.List) == 1 && strings.HasPrefix(group.List[0].Text, "//workflowcheck:ignore") {
+			comments = append(comments, group)
+		}
+	}
+	// Bail if no comments
+	if len(comments) == 0 {
+		return
+	}
+	// Add all present in comment map to ignore map
+	for k := range ast.NewCommentMap(fset, f, comments) {
+		m[k] = struct{}{}
+	}
+}
+
+type collector struct {
+	// Concurrency-safe/immutable fields
+	checker     *Checker
+	pass        *analysis.Pass
+	lookupCache *PackageLookupCache
+	nonDetVars  map[*types.Var]NonDeterminisms
+	ignoreMap   map[ast.Node]struct{}
+
+	funcInfos     map[*types.Func]*funcInfo
+	funcInfosLock sync.Mutex
+}
+
+type funcInfo struct {
+	fn                   *types.Func
+	reasons              NonDeterminisms
+	samePackageCalls     map[*funcInfo]token.Pos
+	samePackageCallsLock sync.Mutex
+	factsApplied         bool
+}
+
+// Concurrency safe
+func (f *funcInfo) addSamePackageCall(callee *funcInfo, pos token.Pos) {
+	// Ignore direct recursive calls
+	if f == callee {
+		return
+	}
+	f.samePackageCallsLock.Lock()
+	defer f.samePackageCallsLock.Unlock()
+	// Only if not already there so we can capture the first token
+	if _, ok := f.samePackageCalls[callee]; !ok {
+		f.samePackageCalls[callee] = pos
+	}
+}
+
+func (c *collector) funcInfo(fn *types.Func) *funcInfo {
+	c.funcInfosLock.Lock()
+	defer c.funcInfosLock.Unlock()
+	info := c.funcInfos[fn]
+	if info == nil {
+		info = &funcInfo{fn: fn, samePackageCalls: map[*funcInfo]token.Pos{}}
+		c.funcInfos[fn] = info
+	}
+	return info
+}
+
+func (c *collector) externalFuncNonDeterminisms(fn *types.Func) NonDeterminisms {
+	return c.lookupCache.PackageNonDeterminisms(fn.Pkg())[fn.FullName()]
+}
+
+func (c *collector) externalVarNonDeterminisms(v *types.Var) NonDeterminisms {
+	return c.lookupCache.PackageNonDeterminisms(v.Pkg())[v.Name()]
+}
+
+func (c *collector) collectFuncInfo(fn *types.Func, decl *ast.FuncDecl) {
+	info := c.funcInfo(fn)
+
+	// If matches a pattern, can eagerly stop here
+	match, ok := c.checker.IdentRefs[fn.FullName()]
```

Comment:
Move `fn.FuncName()` to local var. This is not an inlined getter, but instead [does a bit more](https://github.com/golang/go/blob/go1.17.6/src/go/types/object.go#L469-L490). While I usually don't micro-optimize like this, in this case it's needed.